File: captcha_solver.js
/**
 * Funcaptcha Solver Client
 * =========================
 * Communicates with the local funcaptcha-solver API to solve
 * Arkose Labs (Funcaptcha) challenges for Roblox login.
 */

import axios from 'axios';

// Configuration - replace with your actual config values
const CONFIG = {
    CAPTCHA_SOLVER_URL: process.env.CAPTCHA_SOLVER_URL || 'http://localhost:8003',
    CHROME_VERSION: process.env.CHROME_VERSION || '120.0.6099.109'
};

class CaptchaSolver {
    /**
     * Client for the funcaptcha-solver API.
     * @param {string} solverUrl - The URL of the captcha solver API
     */
    constructor(solverUrl = null) {
        this.solverUrl = solverUrl || CONFIG.CAPTCHA_SOLVER_URL;
        this.session = axios.create({
            timeout: 30000
        });
    }

    /**
     * Solve a Roblox login funcaptcha challenge.
     * @param {string|null} blob - The dxBlob data from the Roblox login challenge response
     * @param {string|null} proxy - Optional proxy string (e.g., "http://user:pass@host:port")
     * @returns {Promise<Object>} Object with 'success', 'token', 'error' keys
     */
    async solveRobloxLogin(blob = null, proxy = null) {
        const taskPayload = {
            preset: 'roblox_login',
            chrome_version: CONFIG.CHROME_VERSION
        };

        if (blob) {
            taskPayload.blob = blob;
        }

        if (proxy) {
            taskPayload.proxy = proxy;
        }

        try {
            // Create the captcha solving task
            const createResp = await this.session.post(
                `${this.solverUrl}/funcaptcha/createTask`,
                taskPayload
            );
            const createData = createResp.data;

            if (!createData.success) {
                return {
                    success: false,
                    token: null,
                    error: `Failed to create task: ${createData.err || 'unknown'}`
                };
            }

            const taskId = createData.task_id;

            // Poll for the result
            const maxWait = 120; // 2 minutes max
            const startTime = Date.now();

            while (Date.now() - startTime < maxWait * 1000) {
                const getResp = await this.session.post(
                    `${this.solverUrl}/funcaptcha/getTask`,
                    { task_id: taskId }
                );
                const result = getResp.data;

                if (result.status === 'completed') {
                    const captchaData = result.captcha || {};
                    if (captchaData.success) {
                        return {
                            success: true,
                            token: captchaData.token,
                            error: null,
                            solveTime: captchaData.procces_time
                        };
                    } else {
                        return {
                            success: false,
                            token: null,
                            error: `Captcha solve failed: ${captchaData.err || 'unknown'}`
                        };
                    }
                }

                // Still processing, wait and retry
                await new Promise(resolve => setTimeout(resolve, 2000));
            }

            return {
                success: false,
                token: null,
                error: 'Captcha solve timed out'
            };

        } catch (error) {
            if (error.code === 'ECONNREFUSED' || error.code === 'ENOTFOUND') {
                return {
                    success: false,
                    token: null,
                    error: 'Cannot connect to captcha solver. Is it running on port 8003?'
                };
            }
            return {
                success: false,
                token: null,
                error: `Captcha solver error: ${error.message}`
            };
        }
    }

    /**
     * Check if the funcaptcha-solver API is reachable.
     * @returns {Promise<boolean>} True if the solver is running
     */
    async isSolverRunning() {
        try {
            await this.session.get(`${this.solverUrl}/`, { timeout: 5000 });
            return true;
        } catch (error) {
            return false;
        }
    }
}

export default CaptchaSolver;

File: roblox_checker/captcha_solver.py
"""
Funcaptcha Solver Client
========================
Communicates with the local funcaptcha-solver API to solve
Arkose Labs (Funcaptcha) challenges for Roblox login.
"""

import time
import requests
from . import config


class CaptchaSolver:
    """Client for the funcaptcha-solver Flask API."""

    def __init__(self, solver_url=None):
        self.solver_url = solver_url or config.CAPTCHA_SOLVER_URL
        self.session = requests.Session()

    def solve_roblox_login(self, blob=None, proxy=None):
        """
        Solve a Roblox login funcaptcha challenge.

        Args:
            blob: The dxBlob data from the Roblox login challenge response.
                  If None, the solver will attempt without blob data.
            proxy: Optional proxy string (e.g., "http://user:pass@host:port")

        Returns:
            dict with 'success', 'token', 'error' keys
        """
        task_payload = {
            "preset": "roblox_login",
            "chrome_version": config.CHROME_VERSION,
        }

        if blob:
            task_payload["blob"] = blob

        if proxy:
            task_payload["proxy"] = proxy

        try:
            # Create the captcha solving task
            create_resp = self.session.post(
                f"{self.solver_url}/funcaptcha/createTask",
                json=task_payload,
                timeout=30,
            )
            create_data = create_resp.json()

            if not create_data.get("success"):
                return {
                    "success": False,
                    "token": None,
                    "error": f"Failed to create task: {create_data.get('err', 'unknown')}",
                }

            task_id = create_data["task_id"]

            # Poll for the result
            max_wait = 120  # 2 minutes max
            start_time = time.time()

            while time.time() - start_time < max_wait:
                get_resp = self.session.post(
                    f"{self.solver_url}/funcaptcha/getTask",
                    json={"task_id": task_id},
                    timeout=30,
                )
                result = get_resp.json()

                if result.get("status") == "completed":
                    captcha_data = result.get("captcha", {})
                    if captcha_data.get("success"):
                        return {
                            "success": True,
                            "token": captcha_data.get("token"),
                            "error": None,
                            "solve_time": captcha_data.get("procces_time"),
                        }
                    else:
                        return {
                            "success": False,
                            "token": None,
                            "error": f"Captcha solve failed: {captcha_data.get('err', 'unknown')}",
                        }

                # Still processing, wait and retry
                time.sleep(2)

            return {
                "success": False,
                "token": None,
                "error": "Captcha solve timed out",
            }

        except requests.exceptions.ConnectionError:
            return {
                "success": False,
                "token": None,
                "error": "Cannot connect to captcha solver. Is it running on port 8003?",
            }
        except Exception as e:
            return {
                "success": False,
                "token": None,
                "error": f"Captcha solver error: {str(e)}",
            }

    def is_solver_running(self):
        """Check if the funcaptcha-solver API is reachable."""
        try:
            resp = self.session.get(f"{self.solver_url}/", timeout=5)
            return True
        except Exception:
            return False